### PR TITLE
feat(speedup): refactor ModulSpeedup V2 using torch.fx graph

### PR DESCRIPTION
**Pull request by Louis-J**
### Description ###

refactor ModulSpeedup V2 using torch.fx graph.
refactor the propagation and make all the intermediate varibales const.
refactor the API to differentiate propagation of different operations.

TODO: pre-set in/out masks, `customized_mask_updaters` doc, `customized_replacers` doc

#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [x] full test - compression

### Checklist ###
  - [x] test case
  - [ ] doc

### CODE REVIEW

File: examples/concretc_tracer/concrete_trace_distilbert.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
+# Copyright (c) Microsoft Corporation.
```
Comment (QuanluZhang) [inline examples/concretc_tracer/concrete_trace_distilbert.py:1]: incorrect folder name "concretc_tracer"

File: examples/concretc_tracer/concrete_trace_yolov3.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        'nan': math.nan,
+        'NoneType': type(None),
+        'torch': torch,
+        'device': torch.device,
+    }
+    exec(code, global_vars, func_dict)
+    return func_dict['forward']
+
+traced_model = concrete_trace(model, {'x': dummy_input})
+recompiled = recompile_from_code(traced_model.code, model.forward.__globals__)
```
Comment (QuanluZhang) [inline examples/concretc_tracer/concrete_trace_yolov3.py:33]: what is `recompiled` used for?

File: examples/model_compress/pruning/activation_pruning_torch_v2.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    evaluator(traced_model)
+
+    # Optimizer used in the pruner might be patched, so recommend to new an optimizer for fine-tuning stage.
+    print('\n' + '=' * 50 + ' START TO FINE TUNE THE MODEL ' + '=' * 50)
+    optimizer, scheduler = optimizer_scheduler_generator(traced_model, _lr=0.01, total_epoch=args.fine_tune_epochs)
+
+    best_acc = 0.0
+    g_epoch = 0
+    for i in range(args.fine_tune_epochs):
+        trainer(traced_model, optimizer, criterion)
```
Comment (QuanluZhang) [inline examples/model_compress/pruning/activation_pruning_torch_v2.py:139]: suggest to rename "trainer" to "train", "evaluator" to "evaluate".

File: nni/compression/pytorch/speedup/v2/container.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+class Slot:
+    def __init__(self) -> None:
+        self.value_0 = None
+        self.value_1 = None
+        self.value_2 = None
+        self.value_3 = None
+        self.mask_0 = None
+        self.mask_1 = None
+        self.mask_2 = None
```
Comment (J-shang) [inline nni/compression/pytorch/speedup/v2/container.py:17]: please add some comments for these `value` and `mask`

File: nni/common/concrete_trace_utils/concrete_proxy.py [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+    map_recursive,
+)
+
+_logger = logging.getLogger(__name__)
+
+@compatibility(is_backward_compatible=True)
+class ConcreteProxy(Proxy):
+    """
+    `ConcreteProxy` is a wrapped proxy carried the real intermediate value.
+    We can use it to trace a more compatibal model, and pass the branches.
```
Comment (QuanluZhang) [inline nni/common/concrete_trace_utils/concrete_proxy.py:43]: -> compatible

File: nni/common/concrete_trace_utils/concrete_proxy.py [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+            # This allows you to create a ConcreteProxy object around a raw Node
+            tracer = et.GraphAppendingConcreteTracer(node.graph)
+        self.tracer = tracer
+        self.value = value
+        self.node = node
+
+    def __repr__(self) -> str:
+        # to detect if in debugging or in code
+        calling_frame_name = inspect.stack()[1][1]
+        if calling_frame_name.endswith('pydevd_exe2.py') or calling_frame_name.endswith('pydevd_safe_repr.py'):
```
Comment (QuanluZhang) [inline nni/common/concrete_trace_utils/concrete_proxy.py:78]: what are "pydevd_exe2.py" and "pydevd_safe_repr.py"?

File: nni/common/concrete_trace_utils/concrete_tracer.py [hunk 7]
```diff
@@ -0,0 +1,1387 @@
...
+    _orig_contains,
+    _orig_index,
+)
+
+HAS_VARSTUFF = inspect.CO_VARARGS | inspect.CO_VARKEYWORDS
+
+@compatibility(is_backward_compatible=True)
+class ConcreteTracer(TracerBase):
+    """
+    A model tracer similar to _symbolic_trace.Tracer, but with concrete execution and real value so we can pass complecate conditions
```
Comment (QuanluZhang) [inline nni/common/concrete_trace_utils/concrete_tracer.py:69]: complex

File: nni/common/concrete_trace_utils/concrete_tracer.py [hunk 8]
```diff
@@ -0,0 +1,1387 @@
...
+                self.c = a + b
+
+        def f(x, y):
+            return leaf_clz(x, y)
+
+        traced_f = concrete_trace(f, (1, 2), autowrap_leaf_class = {
+            leaf_clz: ([], False), **ConcreteTracer.default_autowrap_leaf_class})
+        assert isinstance(traced_f(3, 4), leaf_clz) and traced_f(3, 4).c == 7
+
+    Args:
```
Comment (QuanluZhang) [inline nni/common/concrete_trace_utils/concrete_tracer.py:1325]: seems the docstring does not follow the specified docstring format?

File: nni/common/concrete_trace_utils/operator_patcher.py [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+    _orig_type,
+    _orig_isinstance,
+    _orig_len,
+    _orig_dict,
+    _orig_zip,
+)
+
+_logger = logging.getLogger(__name__)
+
+class TransformerOp(ast.NodeTransformer):
```
Comment (QuanluZhang) [inline nni/common/concrete_trace_utils/operator_patcher.py:28]: "Transformer" in "TransformerOp" is a little misleading, here it means "transformation" rather than "transformer" models, right?

File: nni/common/concrete_trace_utils/concrete_proxy.py (hunk 6 above)
Comment (Louis-J) [reply nni/common/concrete_trace_utils/concrete_proxy.py]: if the func called in 'debug window', we print the 'proxy' info; others we use the normal value info.
we detect the func caller with 'pydevd_exe2.py' and 'pydevd_safe_repr.py'

File: nni/common/concrete_trace_utils/operator_patcher.py (hunk 9 above)
Comment (Louis-J) [reply nni/common/concrete_trace_utils/operator_patcher.py]: yes. it's a 'python func transformer' not something occurred in training process.

File: nni/compression/pytorch/speedup/v2/model_speedup.py [hunk 10]
```diff
@@ -0,0 +1,584 @@
...
+    def mask_merger(self, a, b):
+        if isinstance(a, torch.Tensor) and self.tensor_propagate_check(a):
+            assert isinstance(b, torch.Tensor) and a.shape == b.shape
+            # equals to: torch.maximum(a + b, torch.ones_like(a)) - 1
+            return torch.relu(a + b - 1)
+        else:
+            assert a is None and b is None
+            return None
+
+    def indirect_calc_mask(self, mask, obj):
```
Comment (J-shang) [inline nni/compression/pytorch/speedup/v2/model_speedup.py:260]: who use this

File: nni/compression/pytorch/speedup/v2/model_speedup.py [hunk 11]
```diff
@@ -0,0 +1,584 @@
...
+            assert not isinstance(mask, torch.Tensor)
+
+    # # pass the gradient to the predecessor nodes
+    def indirect_pass_grad(self, slot_val, out):
+        if isinstance(out, torch.Tensor):
+            if self.tensor_propagate_check(out):
+                if slot_val is not None and out.grad is not None:
+                    slot_val += out.grad
+                elif slot_val is None:
+                    slot_val = out.grad
```
Comment (J-shang) [inline nni/compression/pytorch/speedup/v2/model_speedup.py:301]: seems slot_val is a intermediate variables

File: nni/compression/pytorch/speedup/v2/model_speedup.py [hunk 12]
```diff
@@ -0,0 +1,424 @@
...
+            else:
+                return grad
+
+        node_info.output_grad = tree_map_zip(add_grad, node_info.output_grad, outputs)
+
+    def propagate_originally(self):
+        """
+        Propagate normally to get informations of intermediate variables such as shape, dtype of tensors.
+        Default action:
+            execute and store values to slot.value_0(intermediate variables when assigned),
```
Comment (super-dainiu) [inline nni/compression/pytorch/speedup/v2/model_speedup.py:199]: Comments not updated to latest naming.

**Comment by Louis-J** [comment]
@microsoft-github-policy-service agree

**Review by super-dainiu — APPROVED** [review]

File: test/algo/compression/pruning/test_speedup_v2_init_mask.py [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+        x = self.relu1(self.conv1(x))
+        x = self.max_pool1(x)
+        x = self.relu2(self.conv2(x))
+        x = self.max_pool2(x)
+        x = x.view(-1, x.size()[1:].numel())
+        x = self.relu3(self.fc1(x))
+        x = self.fc2(x)
+        return F.log_softmax(x, dim=1)
+
+# class InitMaskTestCase(unittest.TestCase):
```
Comment (QuanluZhang) [inline test/algo/compression/pruning/test_speedup_v2_init_mask.py:87]: why is this part commented out?

File: nni/compression/pytorch/speedup/v2/model_speedup.py [hunk 14]
```diff
@@ -0,0 +1,424 @@
...
+        The path of user provided masks file, or the masks object.
+    map_location
+        The device on which masks are placed, same to map_location in ```torch.load```.
+    batch_dim
+        The index of batch dimension in the dummy_input.
+    batch_size
+        The batch_size coefficient of the sparsity inference.
+        This value is actually used as the batchsize of the dummy_input.
+    customized_mask_updaters
+        ...
```
Comment (QuanluZhang) [inline nni/compression/pytorch/speedup/v2/model_speedup.py:55]: what is the meaning of "..."?

File: nni/compression/pytorch/speedup/v2/model_speedup.py [hunk 15]
```diff
@@ -0,0 +1,424 @@
...
+        #         return b is None
+
+        # for node, masks in node_to_masks.items():
+        #     if len(masks) >= 1:
+        #         mask = masks[0]
+        #         for mask_next in masks[1:]:
+        #             assert check_equal(mask, mask_next), f'preset-masks of "{node}" are not euqal!'
+        #         assert check_valid(self.node_infos[node].output_origin, mask),\
+        #             f'structure of preset-mask and value of slot "{node}" are not euqal!'
+        #         self.node_infos[node].output_masks = mask
```
Comment (QuanluZhang) [inline nni/compression/pytorch/speedup/v2/model_speedup.py:388]: why is this part commented out?

File: test/algo/compression/pruning/test_speedup_v2_init_mask.py (hunk 13 above)
Comment (J-shang) [reply test/algo/compression/pruning/test_speedup_v2_init_mask.py:87]: pre-set in/out masks for speedup has some issues in previous implementation, so I comment the relation code and this test, will add this test back after the pre-set in/out masks feature done.

File: nni/compression/pytorch/speedup/v2/model_speedup.py (hunk 15 above)
Comment (J-shang) [reply nni/compression/pytorch/speedup/v2/model_speedup.py:388]: pre-set in/out masks is not ready, and I don't want to fix it in this pr

File: nni/compression/pytorch/speedup/v2/model_speedup.py (hunk 14 above)
Comment (J-shang) [reply nni/compression/pytorch/speedup/v2/model_speedup.py:55]: Seems a missing TODO, I will add some docstring

**Review by QuanluZhang — APPROVED** [review]

**Review by J-shang — APPROVED** [review]
